google/cloud/firestore_v1/throttle.py

```diff
@@ -0,0 +1,128 @@
+
+import ipdb
+
+
+default_initial_tokens: int = 500
+default_phase_length: int = 60 * 5  # 5 minutes
+microseconds_per_second: int = 1000000
+
+
+class Throttle:
+    """Implements 5/5/5 ramp-up via Token Bucket algorithm."""
+
+    def __init__(
+        self,
+        initial_tokens: Optional[int] = default_initial_tokens,
+        phase_length: Optional[int] = default_phase_length,
+        now_getter: Optional[Callable] = datetime.datetime.utcnow,
+    ):
+        # Tracks the volume of operations during the
+        self._operations_this_phase: int = 0
+
+        # Optional override for telling time. Useful for testing.
+        self._now = now_getter
+
+        self._start: datetime.datetime = self._now()
+        self._last_refill: datetime.datetime = self._start
+
+        # Current number of available operations. Decrements with every
+        # permitted request and refills over time.
+        self._available_tokens: int = initial_tokens
+
+        # Maximum size of the available operations. Can increase by 50%
+        # every [phase_length] number of seconds.
+        self._maximum_tokens: int = self._available_tokens
+
+        # Number of seconds after which the [_maximum_tokens] can increase by 50%.
+        self._phase_length: int = phase_length
+
+        # Tracks how many times the [_maximum_tokens] has increased by 50%.
+        self._phase: int = 0
+
+    def take_token(self) -> bool:
+        """Returns True if an operation is currently permitted or False if not."""
+        self._check_phase()
+        self._refill()
+
+        if self._available_tokens > 0:
+            self._available_tokens -= 1
+            self._operations_this_phase += 1
+            return True
+        return False
+
+    def _check_phase(self):
+        """Increments or decrements [_phase] depending on traffic.
+
+        Every [_phase_length] seconds, if > 50% of available traffic was used
+        during the window, increases [_phase], otherwise, decreases [_phase].
+
+        This is a no-op unless a new [_phase_length] number of seconds since the
+        start was crossed since it was last called.
+        """
+        age: datetime.timedelta = self._now() - self._start
+
+        # Uses integer division to calculate the expected phase. We start in
+        # Phase 0, so until [_phase_length] seconds have passed, this will
+        # not resolve to 1.
+        expected_phase: int = age.seconds // self._phase_length
+
+        # Short-circuit if we are still in the expected phase.
+        if expected_phase == self._phase:
+            return
+
+        operations_last_phase: int = self._operations_this_phase
+        self._operations_this_phase = 0
+
+        previous_phase: int = self._phase
+        self._phase = expected_phase
+
+        # No-op if we did nothing for an entire phase
+        if not operations_last_phase or self._phase > previous_phase + 1:
+            return
+        self._increase_maximum_tokens()
```

Comment:
Might be clearer as:
```suggestion
        if operations_last_phase and self._phase <= previous_phase + 1:
            self._increase_maximum_tokens()
```